materialize-snowflake/snowflake.go

```diff
@@ -465,27 +505,61 @@ func (d *transactor) Store(it *m.StoreIterator) (m.StartCommitFunc, error) {
 			}
 		}
 
-		if err := b.store.stage.start(ctx, d.db); err != nil {
-			return nil, err
+		if !b.streaming {
+			if err := b.store.stage.start(ctx, d.db); err != nil {
+				return nil, err
+			}
 		}
-		converted, err := b.target.ConvertAll(it.Key, it.Values, flowDocument)
-		if err != nil {
+		if converted, err := b.target.ConvertAll(it.Key, it.Values, flowDocument); err != nil {
 			return nil, fmt.Errorf("converting Store: %w", err)
-		}
-		if err = b.store.stage.encodeRow(converted); err != nil {
+		} else if b.streaming {
+			if err := d.streamManager.encodeRow(ctx, it.Binding, converted); err != nil {
+				return nil, fmt.Errorf("encoding Store to stream: %w", err)
+			}
+		} else if err = b.store.stage.encodeRow(converted); err != nil {
 			return nil, fmt.Errorf("encoding Store to scratch file: %w", err)
+		} else {
+			b.store.mergeBounds.NextKey(converted[:len(b.target.Keys)])
 		}
-		b.store.mergeBounds.NextKey(converted[:len(b.target.Keys)])
+	}
+	if it.Err() != nil {
+		return nil, it.Err()
 	}
 
-	// Upload the staged files and build a list of merge and copy into queries that need to be run
-	// to effectively commit the files into destination tables. These queries are stored
-	// in the checkpoint so that if the connector is restarted in middle of a commit
-	// it can run the same queries on the next startup. This is the pattern for
-	// recovery log being authoritative and the connector idempotently applies a commit
-	// These are keyed on the binding table name so that in case of a recovery being necessary
-	// we don't run queries belonging to bindings that have been removed
-	for _, b := range d.bindings {
+	return func(ctx context.Context, runtimeCheckpoint *protocol.Checkpoint) (*pf.ConnectorState, m.OpFuture) {
+		var checkpointJSON, err = d.buildDriverCheckpoint(ctx, runtimeCheckpoint)
```

Comment:
I actually just realized that I was wrong in an earlier commit message about this being able to run in the async part of the transaction - this is the StartCommitFn, which is invoked shortly after `Store` returns. So that part of the justification for the split out `buildDriverCheckpoint` function isn't actually valid. But, the Snowpipe streaming checkpoints do need the runtimeCheckpoint, so I think the restructuring still makes sense.